### PR TITLE
ListOffsets KIP : 396

### PR DESCRIPTION
**Pull request by mahajanadhitya**
… queue implemented

### CODE REVIEW

File: examples/list_offsets.c
Comment (emasab) [inline examples/list_offsets.c:1]: Remove binary file `examples/list_offsets` and add it to `examples/.gitignore`

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:9788]:
```suggestion
const rd_kafka_ListOffsetResultInfo_t **rd_kafka_ListOffsets_result_infos(const rd_kafka_ListOffsets_result_t *result, size_t *cntp);
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:9785]: remove count and use result_infos

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:9782]:
```suggestion
int64_t rd_kafka_ListOffsetResultInfo_timestamp(const rd_kafka_ListOffsetResultInfo_t *result_info);
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:9779]:
```suggestion
const rd_kafka_topic_partition_t *rd_kafka_ListOffsetResultInfo_topic_partition(const rd_kafka_ListOffsetResultInfo_t *result_info);
```

Comment (emasab) [reply src/rdkafka.h:9779]: Documentation

File: src/rdkafka.h
Comment (emasab) [reply src/rdkafka.h:9782]: Documentation

File: src/rdkafka.h
Comment (emasab) [reply src/rdkafka.h:9788]: Documentation

File: src/rdkafka.h
Comment (emasab) [reply src/rdkafka.h:9770]: Documentation

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:5850]: Documentation

File: src/rdkafka.h
Comment (emasab) [reply src/rdkafka.h:6811]: Documentation

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:6999]:
```suggestion
/**
 * @brief Set Isolation Level to an allowed `rd_kafka_IsolationLevel_t` value.
 */
```

File: src/rdkafka.h
Comment (emasab) [reply src/rdkafka.h:9770]: Move ListOffsets functions and enums to a Doxygen section after `Admin API - Configuration`.
```
/**
 * @name Admin API - ListOffsets
 * @{
 */
 ```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8343]: Add param and remark documentation

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8344]:
```suggestion
 * @brief List offset for the specified \p topic_partitions.
 *             This operation enables to find the beginning offset,
 *             end offset as well as the offset matching a timestamp in partitions
 *             or the offset with max timestamp.
```

**Review by emasab — CHANGES_REQUESTED** [review]
First review comments about general checks and `rdkafka.h` additions.

File: src/rdkafka.h
Comment (mahajanadhitya) [reply src/rdkafka.h:6813]: No this is not a new enum will cause code break in our customers if we change the enum name now.

File: src/rdkafka_admin.c
Comment (mahajanadhitya) [reply src/rdkafka_admin.c:4056]: I will do that in the end as I do not have a linux machine so clang does not work and same things for other warnings.

File: src/rdkafka.h
Comment (mahajanadhitya) [reply src/rdkafka.h:949]: Will do via clang at the last.

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:7148]: Should be called `ListOffsetsResultInfo` like in Java, here and everywhere.

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:7146]:
```suggestion
 * Information returned from a ListOffsets call for a specific `rd_kafka_topic_partition_t`.
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:7151]:
```suggestion
 * @brief Returns the topic partition of the passed \p result_info.
```
add `@brief` in other comments too.

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:7157]:
```suggestion
 * Returns the timestamp corresponding to the offset in \p result_info.
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:7131]: Put this Doxygen section after DeleteConsumerGroupOffsets, just before
```
Admin API - User SCRAM credentials
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8811]: Indent it to be below List

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8815]:
```suggestion
 * @param topic_partitions topic_partition_list_t with the partitions and offsets to list.
 *                         Each topic partition offset can be a value of the `rd_kafka_OffsetSpec_t`
 *                         enum or a non-negative value, representing a timestamp, to query for the first
 *                         offset after the given timestamp.
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8821]: Indent (remove one space)

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8810]: Move `rd_kafka_ListOffsets` at the end of its Doxygen section

**Review by emasab — CHANGES_REQUESTED** [review]

File: src/rdkafka_admin.c
Comment (emasab) [reply src/rdkafka_admin.c:4151]: Add `rd_assert(result_element);`, before `result_element->topic_partition->err``

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:7137]: * @enum rd_kafka_OffsetSpec_t
 * @brief Allows to specify the desired offsets when using ListOffsets.

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:7140]: Document the values, for example about this one (javadoc):
```
Used to retrieve the offset with the largest timestamp of a partition as message timestamps can be specified client side this may not match the log end offset returned by SPEC_LATEST.
```

File: src/rdkafka.h
Comment (emasab) [reply src/rdkafka.h]: Missing to move the section

File: src/rdkafka_admin.h
Comment (emasab) [inline src/rdkafka_admin.h:114]:
```suggestion
                                             *   Default Set to RD_KAFKA_ISOLATION_LEVEL_READ_UNCOMMITTED
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8818]:
```suggestion
 *
 * Supported admin options:
 *  - rd_kafka_AdminOptions_set_isolation_level() - default  \c RD_KAFKA_ISOLATION_LEVEL_READ_UNCOMMITTED
 *  - rd_kafka_AdminOptions_set_request_timeout() - default socket.timeout.ms
 *
 * @remark The result event type emitted on the supplied queue is of type
 *         \c RD_KAFKA_EVENT_LISTOFFSETS_RESULT
 */
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:8810]: I took it from Java documentation, but there is a typo there:
```suggestion
 * @brief List offsets for the specified \p topic_partitions.
```

File: src/rdkafka.h
Comment (emasab) [inline src/rdkafka.h:5388]: Add `RD_KAFKA_EVENT_LISTOFFSETS_RESULT` to the documentation of `rd_kafka_event_opaque`

**Review by emasab — CHANGES_REQUESTED** [review]

File: src/rdkafka.h
Comment (mahajanadhitya) [reply src/rdkafka.h:7166]: Since one resultInfo holds the result of one request only it should be ListOffset only instead of ListOffsets also this was discussed in the design as well in our calls.

File: src/rdkafka_request.c
Comment (emasab) [inline src/rdkafka_request.c:614]: `rd_kafka_buf_write_arraycnt_pos`

File: src/rdkafka_request.c
Comment (emasab) [inline src/rdkafka_request.c:638]: for flexver use `rd_kafka_buf_finalize_arraycnt`

File: src/rdkafka_request.c
Comment (emasab) [inline src/rdkafka_request.c:642]: Not needed for top level buffer, as it's written by `rd_kafka_buf_finalize`.
Please remove this final write_tags for `rd_kafka_DescribeUserScramCredentialsRequest` and `rd_kafka_AlterUserScramCredentialsRequest` too. Last skip_tags is also not strictly necessary as it's possible to partially read the buffer.

File: src/rdkafka_request.c
Comment (emasab) [inline src/rdkafka_request.c:487]: use previous `rd_kafka_parse_ListOffsets` instead, upgrading it.

**Review by emasab — CHANGES_REQUESTED** [review]
There are some things to change to keep backward compatibility for consumer `rd_kafka_query_watermark_offsets` and `rd_kafka_offsets_for_times` functions and to upgrade internal consumer calls too.

File: src/rdkafka_request.c
Comment (mahajanadhitya) [reply src/rdkafka_request.c:487]: I find this one much cleaner and better.

File: src/rdkafka_request.c
Comment (mahajanadhitya) [reply src/rdkafka_request.c:565]: That is written very poorly. I do not want to upgrade it. If there are any must needed changes then we can discuss.

**Comment by emasab** [comment]
`rd_kafka_offsets_for_times` needs to be fixed as well but will be in a different PR

File: src/rdkafka.c
Comment (milindl) [inline src/rdkafka.c:3500]: I think we can check here also that if cache is hinted for this topic, and in that case, do nothing as a refresh is already ongoing.

Comment (milindl) [reply src/rdkafka.c:3500]: Same for ListOffsets

Comment (emasab) [reply src/rdkafka.c:3500]: This is already ensured by `rd_kafka_metadata_cache_delete_by_name` because it does a rd_kafka_metadata_cache_find with `valid` equal to 1.

**Review by emasab — APPROVED** [review]
I approve. Given there are my changes too, let's wait for a second approval by @milindl

**Review by milindl — APPROVED** [review]
Thanks for these changes!

**Review by milindl — APPROVED** [review]

**Pull request merged**
